job/dhcp.go

```diff
@@ -97,64 +111,31 @@ func (j Job) areWeProvisioner() bool {
 	return j.hardware.HardwareProvisioner() == j.ProvisionerEngineName()
 }
 
-func (j Job) setPXEFilename(rep *dhcp4.Packet, isPacket, isARM, isUEFI bool) {
-	if j.HardwareState() == "in_use" {
-		if j.InstanceID() == "" {
-			j.Error(errors.New("setPXEFilename called on a job with no instance"))
-			return
-		}
-
-		if j.instance.State != "active" {
-			j.With("hardware.state", j.HardwareState(), "instance.state", j.instance.State).Info("device should NOT be trying to PXE boot")
-			return
+func (j Job) getPXEFilename(arch, firmware string, isOuriPXE bool) string {
+	if !j.isPXEAllowed() {
+		if j.instance != nil && j.instance.State == "active" {
+			// We set a filename because if a machine is actually trying to PXE and nothing is sent it may hang for
+			// a while waiting for any possible ProxyDHCP packets and it would delay booting from disks.
+			// This short cuts all that when we know we want to be booting from disk.
+			return "/pxe-is-not-allowed"
```

Comment:
should this logic be in `isPXEAllowed()`? if we're defining when a machine should pxe or not might be good to have it all in one place. that way when we remove it it's all together! 😁

Reply:
On an unrelated note, I've wondered if rather than returning this /pxe-is-not-allowed we should instead actually return a proper bootloader, that simply boots from HD.   There is an example of this in the netboot.xyz.    I wonder if that would be clearer to the end user.

Reply:
both good points. I'll fix @jacobweinstock's in this PR and open up an issue for your @splaspood

Reply:
oh wait no @jacobweinstock no this doesn't belong in `isPXEAllowed` as that doesn't return a filename and we want to differentiate between "boot to disk by returning a bogus name" vs "really should not be pxe'ing and we're better off causing the boot process to hang, so return nothing"

Reply:
So my first comment didn't capture the root concern that I have. Apologies. What follows here are all suggestions, no worries at all if you disagree. We don't have any written design philosophies so this is all just my opinion. I don't see any reason why this code won't work as-is. I will not hold up this PR if you don't want to make any of the changes I'm suggesting.

My main concern is that I tried to create a mental model/flow chart of the logic for whether to PXE or not and then what to provide if we should PXE but I was unable to do that without having to walk through every line of code starting from `if dhcp.SetupPXE(rep, req) {`.

I found that the naming of functions did not accurately describe their purpose/function. Most of them were multi-purpose functions. With duplicate behavior splattered across many of them. I can provide more details about this if requested.

With that said here are some suggestions just related for `getPXEFilename`. All this with the goals of increasing understandability and maintainability, and decreasing fragility:

- remove `isOuriPXE bool` from the `getPXEFilename `signature
  - if the caller already knows this value is true they should just use `"http://" + conf.PublicFQDN + "/auto.ipxe"`
  - ```go
    if isOuriPXE {
        dhcp.SetFilename(rep, "http://" + conf.PublicFQDN + "/auto.ipxe", conf.PublicIPv4)
    }

    ```
- redefine `getPXEFilename` to just determine a PXE filename based on its inputs. This increases the understandability of the function.
- with a func name of`getPXEFilename` we should define through func comments what this means. something like
```go
// getPXEFilename takes in an architecture and firmware type to determine the filename that should be returned
func (j Job) getPXEFilename(arch, firmware string) string {
``` 
- move and rename `j.isPXEAllowed()` to live in `configureDHCP`. Use the positive if case, something like
```go
if j.bootToDiskImmedately() && j.instance != nil && j.instance.State == "active"{
    dhcp.SetFilename(rep, "/pxe-is-not-allowed", conf.PublicIPv4)
} else {
    dhcp.SetFilename(rep, "", conf.PublicIPv4)
}
```


Ref [Curly's Law](https://blog.codinghorror.com/curlys-law-do-one-thing/#:~:text=Curly's%20Law%20is%20about%20choosing,your%20code%20won't%20do.)

Reply:
sorry, one last thing. You said
> _"boot to disk by returning a bogus name" vs "really should not be pxe'ing and we're better off causing the boot process to hang, so return nothing"_

this business logic feels very opaque. I really would like to see us rethink and reorganize this codebase to be clearer and if possible less opinionated about this type of logic. Anyway, just something to think about.

Reply:
> sorry, one last thing. You said

No worries, keep em coming!

> 
> > _"boot to disk by returning a bogus name" vs "really should not be pxe'ing and we're better off causing the boot process to hang, so return nothing"_
> 
> this business logic feels very opaque. I really would like to see us rethink and reorganize this codebase to be clearer and if possible less opinionated about this type of logic. Anyway, just something to think about.

Agree, I hope by removing cacher stuff we can then make progress on removing packet/EM only things (phone-home, password callback, ...) and then have a much smaller code base where we can see all this opinionated business logic that we can hash out if they make sense or not and clean up if they do.